### PR TITLE
Drop Delta cache without metastore communication

### PR DESCRIPTION
**Pull request by findepi**
Refactor Delta flush_metadata_cache procedure so that it doesn't have to
rely on external state (the metastore) to drop the caches.

Follow-up to https://github.com/trinodb/trino/pull/17174#discussion_r1174008917

### CODE REVIEW

**Comment by findepi** [comment]
Currently based on https://github.com/trinodb/trino/pull/17092

**Comment by findepi** [comment]
> Currently based on #17092

rebased

File: testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportBadTestAnnotations.java [hunk 1]
```diff
@@ -64,13 +66,59 @@ private void reportUnannotatedTestMethods(ITestClass testClass)
...
+        for (Class<?> clazz = realClass; clazz != null; clazz = clazz.getSuperclass()) {
+            org.testng.annotations.Test testAnnotation = clazz.getAnnotation(org.testng.annotations.Test.class);
+            if (testAnnotation != null && isAllDefaults(testAnnotation, org.testng.annotations.Test.class)) {
+                return Optional.of(clazz);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static <T extends Annotation> boolean isAllDefaults(T annotationInstance, Class<T> annotationClass)
```
Comment (alexjo2144) [inline testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportBadTestAnnotations.java:102]: `isAllDefaults` is always called with the same second argument. Could be removed, yeah?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/statistics/CachingExtendedStatisticsAccess.java [hunk 2]
```diff
@@ -68,32 +70,47 @@ public Optional<ExtendedStatistics> readExtendedStatistics(ConnectorSession sess
...
 
-    public void invalidateCache(String tableLocation)
+    // for explicit cache invalidation
+    public void invalidateCache(SchemaTableName schemaTableName, Optional<String> tableLocation)
     {
-        // for explicit cache invalidation
-        cache.invalidate(tableLocation);
+        requireNonNull(schemaTableName, "schemaTableName is null");
+        if (tableLocation.isPresent()) {
+            cache.invalidate(new CacheKey(schemaTableName, tableLocation.get()));
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/statistics/CachingExtendedStatisticsAccess.java:96]: I think you want to invalidate based on location and name each separately here, even if the name is present. It is probably rare but an external system could drop and recreate a table using the same location but a different name, or same name in a different location.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java [hunk 3]
```diff
@@ -182,10 +185,18 @@ public void flushCache()
...
-    public void invalidateCaches(String tableLocation)
+    public void invalidateCache(SchemaTableName schemaTableName, Optional<String> tableLocation)
     {
-        tableSnapshots.invalidate(tableLocation);
-        activeDataFileCache.invalidate(tableLocation);
+        requireNonNull(schemaTableName, "schemaTableName is null");
+        if (tableLocation.isPresent()) {
+            CacheKey cacheKey = new CacheKey(schemaTableName, tableLocation.get());
+            tableSnapshots.invalidate(cacheKey);
+            activeDataFileCache.invalidate(cacheKey);
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java:194]: Potentially same here, invalidating based on name and location separately?

**Review by findinpath — APPROVED** [review]

**Review by findinpath — APPROVED** [review]

**Review by findinpath — APPROVED** [review]

**Review by findinpath — APPROVED** [review]

**Review by findinpath — APPROVED** [review]

**Review by findinpath — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 4]
```diff
@@ -958,7 +958,7 @@ private void invalidateTableStatisticsCache(String databaseName, String tableNam
     {
         tableStatisticsCache.asMap().keySet().stream()
                 .filter(table -> table.getDatabaseName().equals(databaseName) && table.getTableName().equals(tableName))
-                .forEach(tableCache::invalidate);
+                .forEach(tableStatisticsCache::invalidate);
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java]: Can you test this via `TestCachingHiveMetastore`?

**Comment by findepi** [comment]
Ci 
https://github.com/trinodb/trino/issues/17158

**Comment by findepi** [comment]
(just rebased)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/statistics/CachingExtendedStatisticsAccess.java (hunk 2 above)
Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/statistics/CachingExtendedStatisticsAccess.java:96]: for this to matter, one location would need to be used interchangibly with different table names, right?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeInsertTableHandle.java [hunk 5]
```diff
@@ -65,6 +67,12 @@ public String getTableName()
         return tableName;
     }
 
+    @JsonIgnore
+    public SchemaTableName getSchemaTableName()
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeInsertTableHandle.java]: good idea

**Comment by findepi** [comment]
AC

**Review by losipiuk — APPROVED** [review]
skimmed

**Comment by findepi** [comment]
rebased to resolve conflicts

**Comment by findepi** [comment]
(rebased)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/statistics/CachingExtendedStatisticsAccess.java (hunk 2 above)
Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/statistics/CachingExtendedStatisticsAccess.java:95]: applied. note that this way of invalidation will be slower (linear search)

**Comment by findepi** [comment]
AC

**Pull request merged**